src/python/pants/util/objects.py

```diff
@@ -349,28 +321,40 @@ def resolve_for_enum_variant(self, mapping):
         raise self.make_type_error(
           "pattern matching must have exactly the keys {} (was: {})"
           .format(self._allowed_values, list(keys)))
-      match_for_variant = mapping[getattr(self, field_name)]
+      match_for_variant = mapping[self.underlying()]
       return match_for_variant
 
+    def underlying(self):
+      """Get the element of `all_values` corresponding to this enum instance.
+
+      This should be used only for generating option values in tests. In general, it is less
+      error-prone to deal with enum objects directly.
+      """
+      return getattr(self, field_name)
+
     @classmethod
     def iterate_enum_variants(cls):
       """Iterate over all instances of this enum, in the declared order.
 
       NB: This method is exposed for testing enum variants easily. resolve_for_enum_variant() should
       be used for performing conditional logic based on an enum instance's value.
       """
-      # TODO(#7232): use this method to register attributes on the generated type object for each of
-      # the singletons!
       return cls._singletons.values()
 
+  # Python requires creating an explicit closure to save the value on each loop iteration.
+  accessor_generator = lambda case: lambda cls: cls(case)
```

Comment:
I don't think you need the inner `lambda cls` here, as `cls` doesn't vary across loop iterations

Reply:
This seems correct, I just picked what seemed to work because python frustrates me :). Will fix.

Reply:
Oh, the `lambda cls` was just because we needed to have a function to pass to `classproperty` and this seemed pretty concise.